### ISSUE TITLE
PROPOSAL / API change: proper node UUID's instead of using bindHttp as node identifier

### ISSUE THREAD

**Issue by steveyen** [couchbaselabs/cbgt#25]
As part of the cbft integration project with Couchbase Server (ns-server), I had a discussion with @aartamonau on node identities, and he _strongly_ recommended and counseled that cbgt should instead use proper, opaque identifiers for node identity.

In contrast, today, cbgt overloads the the "bindHttp" command-line flag value for two uses: for node identity and for network addressing.

The current bindHttp approach has ugliness with node IP address changes (e.g., DHCP with worker moving from home to cafe to work).  And, there were plenty of battle stories from ns-server which uses non-UUID, non-unique identifiers, such as around upgrades; and battle stories around failover'ed or removed nodes that are re-added back into the cluster; and battle stories around nodes of the wrong version re-entering a cluster.  In those battle stories, the non-unique node identifiers (like n_1@192.0.0.10) wasn't strong enough for the system to detect issues on these edge cases.

cbgt nodes currently do have a Manager UUID, but that isn't used regularly throughout cbgt as a straight identity value all the time...
- https://github.com/couchbaselabs/cbgt/blob/master/defs.go#L112
- https://github.com/couchbaselabs/cbgt/blob/master/manager.go#L136

The proposal is to embrace the NodeDef / Manager UUID as the node's true identifier, and introduce some new mapping or lookup-table (in the Cfg) that allows one to find the network-address(es) of a node from its UUID.

Some other impact, changes, and non-changes...
- Cfg contents would change, in an incompatible way, where the nodeDefs mapping would be changed to be keyed by UUID instead of bindHttp.

https://github.com/couchbaselabs/cbgt/blob/master/defs.go#L105
- a new node UUID -> networkAddress(es) lookup mapping would be also stored in the Cfg
- The cbft IndexAlias / RemoteClient codepaths would have to use this UUID -> networkAddress(es) lookup mapping as they scatter/gather queries across nodes.

https://github.com/couchbaselabs/cbft/blob/master/pindex_bleve.go#L831
- the cbgt UI would change on its "Nodes" screen.
- command-line parameters (-bindHttp) would not likely have to change.

**Comment by steveyen** [comment]
cbgt's Manager API would likely also _not_ have to change.

but, cbgt's CfgGetNodeDefs() API and NodeDefs json struct API's would change in an incompatible, breaking way.

**Comment by mschoch** [comment]
When implementing the stats code, I have code that walks those node defs to extract node specific information.  It sounds like this change is going to make my life harder...

**Comment by tleyden** [comment]
> command-line parameters (-bindHttp) would not likely have to change.

Wouldn't that command line parameter be removed with this change?

**Comment by mschoch** [comment]
Why would it be removed, shouldn't it still serve its original purpose, let you control what interface(s) the server binds to?

**Comment by tleyden** [comment]
Oh, I forgot about that other purpose.  Per @steveyen's [comment](https://github.com/couchbaselabs/cbgt/issues/1#issuecomment-121714106):

> If you don't use the REST facilities (which actually listen on that bindHttp / see http://godoc.org/github.com/couchbaselabs/cbgt/rest#InitRESTRouter), then I guess bindHttp doesn't need to be actually "routable" and only needs to be unique for each node (so, it might work for bindHttp to be completely artificial). 

We are not yet using the REST facilities in Sync Gateway, so we are just making sure `bindHttp` is unique per-node.  Eg, I've been using "sg1" and "sg2" in my testing.

If and when we do add the CBGT REST facilities, we'd probably just add a new endpoint to our existing router, and reuse the bind interface/port that is already specified in the Sync Gateway config -- that config defaults to 0.0.0.0:4985, so it's probably not useful as a unique identifier for a node.

**Comment by mschoch** [comment]
Oh good point.  Since cbgt is a library, should it even be doing HTTP itself?  Wouldn't it make more sense that it either provide(and or register) handlers, but let the application take responsibility for setting that up (if appropriate)?

**Comment by steveyen** [comment]
> Oh good point. Since cbgt is a library, should it even be doing HTTP itself?

Hi Marty,
Some quick notes on the two sides of HTTP, client & server, and cbgt...

On the client-side, currently, cbgt doesn't know how to be an HTTP/REST client.  For example, for cbft (couchbase full text server), all the HTTP/REST client stuff lives in cbft instead of cbgt.

In the future, though, I've been wondering if there might be future approaches to pull the HTTP/REST stuff down a level and make it reusable across different PIndex implementations.  No brilliant a-ha moments on that yet, though, except that it might be more worthwhile if we ever get around back to next-gen "greenstack" networking proposals

> Wouldn't it make more sense that it either provide(and or register) handlers, but let the application take responsibility for setting that up (if appropriate)?

On the server-side, yeah, it's kinda like you're describing.  From the bottom's up...

cbgt - there's the core cbgt library that doesn't have any REST/HTTP stuff at all, but only has the generic Cfg stuff and generic manager/planner/janitor stuff.

cbgt/rest - then there's a higher-level library, cbgt/rest, that apps can optionally use to have some pre-built REST/HTTP handlers, or if app writers just want to pass in a gorilla.mux.Router that cbgt/rest can populate with all the handlers & routes, that's available too.  See...

http://godoc.org/github.com/couchbaselabs/cbgt/rest
http://godoc.org/github.com/couchbaselabs/cbgt/rest#InitRESTRouter

The cbgt/rest library also has web admin UI static resources that apps can optionally use.  That is, the framework of all the angularJS lives under cbgt/rest/static, where it's conveniently (?) bundled up for optional app usage via a bindata_assetfs.go approach.

Lastly...

cbgt/cmd - then there's another higher-level library, cbgt/cmd, that can apps can optionally use to get some help with with cmd-line flags parsing and initialization.

So, a little bit of a'la carte, where the carte is a bit underdocumented.

**Comment by mschoch** [comment]
OK, I'm confused, so I'll just leave it to you two to figure out.

**Comment by steveyen** [comment]
The changes on this have been pushed to cbgt...

commit 4bddecfd

And these are some related web admin UI changes to cbgt...

commit 09ee7707
commit b53b2c7
commit 430baef1

Interestingly, looks like there were no changes needed to cbft or to cbmirror.

Sync-gateway, however, might need changes if it accesses the NodeDefs data structure (e.g., if the app invokes the cbgt.CfgGetNodeDefs() API or similar).

In particular, the key of the NodeDefs map[string]*NodeDef has changed from a hostPort string to a UUID string.  The compiler, of course, thinks its just strings either way, so beware that the compiler doesn't save us on this kind of breaking change!

**Comment by mschoch** [comment]
Hi Steve,

I'm reviewing the impact to the stats for ns_server and would like you to take a look as well.

We get the NodeDefs here: https://github.com/couchbaselabs/cbft/blob/master/ns_server.go#L307

pass them to a function call HostsForIndex here: https://github.com/couchbaselabs/cbft/blob/master/ns_server.go#L347

Then in this function we do the following: https://github.com/couchbaselabs/cbft/blob/master/ns_server.go#L266-L295
1. collect node UUIDs participating in the index (from planPIndexes not NodeDefs)
2. walk the nodeDefs, looking at values, ignoring keys, if the nodeUUID is one of the ones we collected in step 1, then remember its nodeExtras value, keyed by the UUID, but again thats just for uniqueness in the map, we don't care what that is.
3. flatten the map of nodeExtras we built into a slice of strings and sort it

It _seems_ to me like all of this should still work, because the host/port info we share comes out of nodeExtras and _not_ out of the nodes UUID or bind info.

So, based on that I think its worth me giving it a try...

BUT, I have a new concern.  The value I'm passing into nodeExtras at cbft startup is _supposed_ to match the identifier for the node by ns_server.  So, I'm wondering, based on this talk about ns_server having some late binding, and changing the name, that this may in fact be broken.  Is there any documentation where I can read more about ns_server changing identification of a node?  I had to add this code once I ran a cluster because it seems like one node continues to identify itself as 127.0.0.1 and the other gets renamed to 192.x.x.x.

**Comment by steveyen** [comment]
On docs on the ip address rename, the proposed design got simpler due to using proper node UUID's...

https://github.com/couchbaselabs/cbft/blob/master/DESIGN.md#handling-ip-address-changes

Of note, I don't think we've yet built what's proposed, where where ns-server restarts cbft with the proper bindHttp value.

On the ns-server side, there's some documentation on the node joining sequence (where IP address changes are a mention-in-passing part of that), so really this pointer is about pointers to further places in the code...

https://github.com/couchbase/ns_server/blob/master/doc/cluster-add-sequence.txt

**Comment by steveyen** [comment]
Also, my reading of the nsstats code was the same as yours.  That is, I was pleasantly surprised with the takeaway that "Hmmm, I don't think this code needs changing at all for this node UUID code change to work"

**Comment by mschoch** [comment]
Ah OK, let me try again.  My new concern is not with cbft's handling of the name change.  My concern is that I'm passing some data into CBFT on startup.  But what if ns_server decides to change the name after that point, and now the "name" I passed into cbft has become stale.

**Comment by steveyen** [comment]
Interesting... Here's my takeaway notes after a look...

1) That extra name gets passed to cbft through the cmd-line, via the -extras JSON.

2) Then saved into the Cfg.

3) Then, later, the nssstats code reads the Cfg.

4) And, nsstats uses a uuid-based match to get the extras JSON.

So, my reading of the code makes me think it should just work (although I can see an optimization at step 4 on perhaps using a straight hash lookup instead of a loop).

I'm thinking that even if ns-server restarts cbft with a changed -extras cmd-line flag parameter (per the DESIGN.md proposal on ip-address changes) and goes through the 1-2-3-4 steps again, it'll be fine.

EXCEPT, at step 2.5, what pops to mind is if the Cfg doesn't get updated or loses its mind (or is eventually consistent instead of consistent), the the wrong extras will come out.

Well, I'd agree that can happen.  But, I'm also thinking that could have happened before in the previous "pre-node-UUID" world, and we just need to wait for metakv / Cfg to converge.

**Comment by mschoch** [comment]
OK, I think where I'm not confident is that you say "even if ns-server restarts cbft with a changed -extras cmd-line flag parameter"...

What reason do we have to think it would do that?  My fear is that it changes the name, and doesn't restart cbft, and we wrongly operate with the old value.

**Comment by mschoch** [comment]
Also, I agree this isn't a new problem introduced by this change.  It was always a problem, but new information about nodes changing names surfaced in the discussion.

**Comment by steveyen** [comment]
》OK, I think where I'm not confident is that you say "even if ns-server restarts cbft with a changed -extras cmd-line flag parameter"...
》 What reason do we have to think it would do that?

No reason - afaik, ns-server doesn't restart cbft on node name changes... yet.  It's future, to-do engineering work...

> Of note, I don't think we've yet built what's proposed, where where ns-server restarts cbft with the proper bindHttp value.

...which is why I'm often pleasantly surprised when I hear that clustering "works".  (Or, based on this morning's email threads, I hear it no longer works.)  The only way I figure it could have "worked" was under a cluster-run scenario where everything (including clients / browser) are running under 127.0.0.1, same machine.  But, at this stage, if you tried to do this across multiple real machines, without handling the (now obsoleted) proposed, convoluted ipaddress renaming or (now implemented) node-UUID or (not yet implemented) cbft restarts, it'd be a "huh, no way that should work yet" mystery.

**Comment by mschoch** [comment]
Hmmn, I guess I don't understand everything thats involved.  I would expect it to just work.  Right now neither cbft node even sees any other nodes registered in its config.
